plutus-ledger/src/Ledger/Typed/Scripts/StakeValidators.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE KindSignatures    #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE ViewPatterns      #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-strictness #-}
+{-# OPTIONS_GHC -Wno-simplifiable-class-constraints #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+module Ledger.Typed.Scripts.StakeValidators (
+    WrappedStakeValidatorType
+    , wrapStakeValidator
+    , mkForwardingStakeValidator
+    , forwardToValidator
+    , Any
+    ) where
+
+import           PlutusTx
+import           PlutusTx.Prelude
+
+import           Plutus.V1.Ledger.Address    (Address (..))
+import           Plutus.V1.Ledger.Contexts   (ScriptContext (..), ScriptPurpose (..), TxInfo (..))
+import qualified Plutus.V1.Ledger.Contexts   as Validation
+import           Plutus.V1.Ledger.Credential (Credential (..))
+import           Plutus.V1.Ledger.Scripts
+import           Plutus.V1.Ledger.Tx         (TxOut (..))
+
+import           Ledger.Typed.TypeUtils
+
+type WrappedStakeValidatorType = BuiltinData -> BuiltinData -> ()
+
+-- TODO: we should add a TypedStakeValidator interface here
+
+{-# INLINABLE wrapStakeValidator #-}
+wrapStakeValidator
+    :: UnsafeFromData r
+    => (r -> Validation.ScriptContext -> Bool)
+    -> WrappedStakeValidatorType
+-- We can use unsafeFromBuiltinData here as we would fail immediately anyway if parsing failed
+wrapStakeValidator f r p = check $ f (unsafeFromBuiltinData r) (unsafeFromBuiltinData p)
+
+-- | A stake validator that checks whether the validator script was run
+--   in the right transaction.
+mkForwardingStakeValidator :: ValidatorHash -> StakeValidator
+mkForwardingStakeValidator vshsh =
+    mkStakeValidatorScript
+    $ $$(PlutusTx.compile [|| \(hsh :: ValidatorHash) -> wrapStakeValidator (forwardToValidator hsh) ||])
+       `PlutusTx.applyCode` PlutusTx.liftCode vshsh
+
+{-# INLINABLE forwardToValidator #-}
+forwardToValidator :: ValidatorHash -> () -> ScriptContext -> Bool
+forwardToValidator h _ ScriptContext{scriptContextTxInfo=TxInfo{txInfoInputs}, scriptContextPurpose} =
+    case scriptContextPurpose of
+        Minting _ -> False
+        Spending _ -> False
+        _ ->
```

Comment:
suspicious, better to list them! Also, I'm not sure if we really need this check (nor in the minting policy case).

Reply:
I didn't want code duplication but I wasn't thinking straight.